### PR TITLE
Add `rustc_on_unimplemented` to `Sum` and `Product` trait.

### PR DESCRIPTION
**Pull request by aDotInTheVoid**
Helps with #105184, but I don't think it fully fixes it.

### CODE REVIEW

File: library/core/src/iter/traits/accum.rs [hunk 1]
```diff
@@ -27,6 +31,10 @@ pub trait Sum<A = Self>: Sized {
 /// [`product()`]: Iterator::product
 /// [`FromIterator`]: iter::FromIterator
 #[stable(feature = "iter_arith_traits", since = "1.12.0")]
+#[rustc_on_unimplemented(
```
Comment (aDotInTheVoid) [inline library/core/src/iter/traits/accum.rs:34]: I'm not super happy with this wording, but can't think of something better.

File: library/core/src/iter/traits/accum.rs [hunk 2]
```diff
@@ -27,6 +31,10 @@ pub trait Sum<A = Self>: Sized {
...
+    message = "a value of type `{Self}` cannot be made by for the product of an iterator with elements of type `{A}`",
+    label = "value of type `{Self}` cannot be made with the product a `std::iter::Iterator<Item={A}>`"
```
Comment (compiler-errors) [inline library/core/src/iter/traits/accum.rs:36]: This isn't grammatical I think, though I don't know if there's an analogous verb to "sum" for "multiplying all together"... Maybe "multiplying" works though.

Comment (estebank) [reply library/core/src/iter/traits/accum.rs:36]: I think that is a reasonable wording here.

**Review by estebank — APPROVED** [review]

File: library/core/src/iter/traits/accum.rs (hunk 2 above)
Comment (estebank) [reply library/core/src/iter/traits/accum.rs:36]: @aDotInTheVoid I can r+ once you've changed the wording here to what you drafted in the comment :)

Comment (aDotInTheVoid) [reply library/core/src/iter/traits/accum.rs]: Updated

**Comment by aDotInTheVoid** [comment]
Rebased onto latest master. I wounder why RLA didn't triger.

**Comment by estebank** [comment]
You created a new test whose output changed since you created it. It wouldn't have been caught until CI caught it during rollup. I just knew about it because I caused the change :)

**Pull request merged**